syntax = "proto3";

package dfs_service;

service DFSService {

    // Add your service calls here

    // 1. REQUIRED (Parts 1 & 2): A method to store files on the server
    rpc Store (stream FileChunk) returns (StoreResponse);

    // 2. REQUIRED (Parts 1 & 2): A method to fetch files from the server
    rpc Fetch (FetchRequest) returns (stream FileChunk);

    // 3. REQUIRED (Parts 1 & 2): A method to delete files from the server
    rpc Delete (DeleteRequest) returns (DeleteResponse);

    // 4. REQUIRED (Parts 1 & 2): A method to list all files on the server
    rpc List (ListRequest) returns (ListResponse);

    // 5. REQUIRED (Parts 1 & 2): A method to get the status of a file on the server

    // 7. Any other methods you deem necessary to complete the tasks of this assignment


}

// Add your message types here
message FileChunk {
    string file_name = 1;
    bytes data = 2;
}

message StoreResponse {
    bool ok = 1;
    string message = 2;
}

message FetchRequest {
    string file_name = 1;
}

message DeleteRequest {
    string file_name = 1;
}

message DeleteResponse {
    bool ok = 1;
    string message = 2;
}

message ListRequest {}

message ListResponse {
    map<string, int64> file_info=1; // filename -> unix epoch millis
}